File: kotlin-spring-data-jpa/src/test/kotlin/com/kotlin_spring_data_jpa/kotlin_spring_data_jpa/controllers/BeerControllerTest.java
package com.kotlin_spring_data_jpa.kotlin_spring_data_jpa.controllers;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.kotlin_spring_data_jpa.kotlin_spring_data_jpa.models.BeerDTO;
import com.kotlin_spring_data_jpa.kotlin_spring_data_jpa.services.BeerService;
import com.kotlin_spring_data_jpa.kotlin_spring_data_jpa.services.BeerServiceImpl;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.ArgumentCaptor;
import org.mockito.Captor;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.MockMvc;

import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;
import static org.hamcrest.core.Is.is;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.BDDMockito.given;
import static org.mockito.Mockito.verify;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;

@WebMvcTest(BeerController.class)
class BeerControllerTest {

    @Autowired
    MockMvc mockMvc;

    @Autowired
    ObjectMapper objectMapper;

    @MockBean
    BeerService beerService;

    @Captor
    ArgumentCaptor<UUID> uuidArgumentCaptor;

    @Captor
    ArgumentCaptor<BeerDTO> beerCaptor;

    BeerServiceImpl beerServiceImpl;


    @BeforeEach
    void setUp() {
        beerServiceImpl = new BeerServiceImpl();
    }

    @Test
    void testGetAllBeers() throws Exception {
        given(beerService.listBeer()).willReturn(beerServiceImpl.listBeer());

        mockMvc.perform(get(BeerController.BASE_BEER_PATH)
                        .accept(MediaType.APPLICATION_JSON))
                .andExpect(status().isOk())
                .andExpect(content().contentType(MediaType.APPLICATION_JSON))
                .andExpect(jsonPath("$.length()", is(beerServiceImpl.listBeer().size())));
    }

    @Test
    void testGetBeerById() throws Exception {
        BeerDTO testBeerDTO = beerServiceImpl.listBeer().getFirst();

        given(beerService.getBeerById(testBeerDTO.getId())).willReturn(testBeerDTO);

        mockMvc.perform(get(BeerController.BEER_PATH_WITH_ID, testBeerDTO.getId())
                        .accept(MediaType.APPLICATION_JSON))
                .andExpect(status().isOk())
                .andExpect(content().contentType(MediaType.APPLICATION_JSON))
                .andExpect(jsonPath("$.id", is(testBeerDTO.getId().toString())))
                .andExpect(jsonPath("$.name", is(testBeerDTO.getName())));
    }

    @Test
    void testGetBeerByIdNotFound() throws Exception {
        given(beerService.getBeerById(any(UUID.class))).willReturn(null);

        mockMvc.perform(get(BeerController.BEER_PATH_WITH_ID, UUID.randomUUID()))
                .andExpect(status().isNotFound());
    }

    @Test
    void testCreteNewBeer() throws Exception {
        BeerDTO testBeerDTO = beerServiceImpl.listBeer().getFirst();

        testBeerDTO.setVersion(null);
        testBeerDTO.setId(null);

        given(beerService.save(any(BeerDTO.class))).willReturn(beerServiceImpl.listBeer().getFirst());

        mockMvc.perform(post(BeerController.BASE_BEER_PATH)
                        .accept(MediaType.APPLICATION_JSON)
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(objectMapper.writeValueAsString(testBeerDTO)))
                .andExpect(status().isCreated())
                .andExpect(header().exists("Location"));
    }

    @Test
    void testUpdateBeer() throws Exception {
        BeerDTO testBeerDTO = beerServiceImpl.listBeer().getFirst();

        mockMvc.perform(put(BeerController.BEER_PATH_WITH_ID, testBeerDTO.getId())
                        .accept(MediaType.APPLICATION_JSON)
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(objectMapper.writeValueAsString(testBeerDTO)))
                .andExpect(status().isNoContent());

        verify(beerService).updateById(any(UUID.class), any(BeerDTO.class));
    }

    @Test
    void testDeleteBeer() throws Exception {
        BeerDTO testBeerDTO = beerServiceImpl.listBeer().getFirst();

        assertNotNull(testBeerDTO.getId(), "The beer ID should not be null");

        mockMvc.perform(delete(BeerController.BEER_PATH_WITH_ID, testBeerDTO.getId())
                        .accept(MediaType.APPLICATION_JSON))
                .andExpect(status().isNoContent());

        verify(beerService).deleteById(uuidArgumentCaptor.capture());

        assertThat(uuidArgumentCaptor.getValue()).isEqualTo(testBeerDTO.getId());
    }

    @Test
    void testPatchBeer() throws Exception {
        BeerDTO testBeerDTO = beerServiceImpl.listBeer().getFirst();

        Map<String, Object> beerMap = new HashMap<>();
        beerMap.put("name", "New Name");

        mockMvc.perform(patch(BeerController.BEER_PATH_WITH_ID, testBeerDTO.getId())
                        .accept(MediaType.APPLICATION_JSON)
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(objectMapper.writeValueAsString(beerMap))
                )
                .andExpect(status().isNoContent());

        verify(beerService).patchById(uuidArgumentCaptor.capture(), beerCaptor.capture());
        assertThat(uuidArgumentCaptor.getValue()).isEqualTo(testBeerDTO.getId());
        assertThat(beerCaptor.getValue().getName()).isEqualTo(beerMap.get("name"));
    }
}

File: kotlin-spring-r2dbc/src/test/kotlin/com/kotlin_spring_r2dbc/kotlin_spring_r2dbc/controller/BeerControllerTest.kt
package com.kotlin_spring_r2dbc.kotlin_spring_r2dbc.controller

import com.kotlin_spring_r2dbc.kotlin_spring_r2dbc.model.BeerDTO
import com.kotlin_spring_r2dbc.kotlin_spring_r2dbc.repositories.BeerRepositoryTest
import org.junit.jupiter.api.MethodOrderer
import org.junit.jupiter.api.Order
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestMethodOrder
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.reactive.AutoConfigureWebTestClient
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.test.web.reactive.server.WebTestClient
import reactor.core.publisher.Mono

@TestMethodOrder(MethodOrderer.OrderAnnotation::class)
@SpringBootTest
@AutoConfigureWebTestClient
class BeerControllerTest {
    @Autowired
    lateinit var webTestClient: WebTestClient

    @Test
    @Order(1)
    fun testListBeers() {
        webTestClient
            .get()
            .uri(BeerController.BEER_PATH)
            .exchange()
            .expectStatus().isOk
            .expectHeader().valueEquals("Content-type", "application/json")
            .expectBody().jsonPath("$.size()").isEqualTo(3)
    }

    @Test
    @Order(2)
    fun testGetBeerById() {
        webTestClient
            .get()
            .uri(BeerController.BEER_PATH_ID, 1)
            .exchange()
            .expectStatus().isOk
            .expectHeader().valueEquals("Content-type", "application/json")
            .expectBody(BeerDTO::class.java)
    }

    @Test
    fun testGetBeerByIdNotFound() {
        webTestClient
            .get()
            .uri(BeerController.BEER_PATH_ID, 1111111)
            .exchange()
            .expectStatus().isNotFound
    }

    @Test
    @Order(3)
    fun testUpdateBeer() {
        webTestClient.put()
            .uri(BeerController.BEER_PATH_ID, 1)
            .body(Mono.just(BeerRepositoryTest.getTestBeer()), BeerDTO::class.java)
            .exchange()
            .expectStatus().isNoContent
    }

    @Test
    @Order(4)
    fun testUpdateBeerBad() {
        val testBeer = BeerRepositoryTest.getTestBeer()
        testBeer.style = ""

        webTestClient.put()
            .uri(BeerController.BEER_PATH_ID, 1)
            .body(Mono.just(testBeer), BeerDTO::class.java)
            .exchange()
            .expectStatus().isBadRequest
    }

    @Test
    fun testUpdateBeerNotFound() {
        val testBeer = BeerRepositoryTest.getTestBeer()

        webTestClient.put()
            .uri(BeerController.BEER_PATH_ID, 111)
            .body(Mono.just(testBeer), BeerDTO::class.java)
            .exchange()
            .expectStatus().isNotFound
    }

    @Test
    fun testPatchBeerNotFound() {
        val testBeer = BeerRepositoryTest.getTestBeer()

        webTestClient.patch()
            .uri(BeerController.BEER_PATH_ID, 111)
            .body(Mono.just(testBeer), BeerDTO::class.java)
            .exchange()
            .expectStatus().isNotFound
    }

    @Test
    fun testCreateBeer() {
        webTestClient
            .post()
            .uri(BeerController.BEER_PATH)
            .body(Mono.just(BeerRepositoryTest.getTestBeer()), BeerDTO::class.java)
            .header("Content-type", "application/json")
            .exchange()
            .expectStatus().isCreated
            .expectHeader().location("http://localhost:8080/api/v2/beers/4")
    }

    @Test
    fun testCreateBeerBad() {
        val testBeer = BeerRepositoryTest.getTestBeer()
        testBeer.name = ""


        webTestClient
            .post()
            .uri(BeerController.BEER_PATH)
            .body(Mono.just(testBeer), BeerDTO::class.java)
            .header("Content-type", "application/json")
            .exchange()
            .expectStatus().isBadRequest
    }

    @Test
    @Order(999)
    fun testDeleteBeer() {
        webTestClient.delete()
            .uri(BeerController.BEER_PATH_ID, 1)
            .exchange()
            .expectStatus().isNoContent
    }
}